cli_tools/common/utils/storage/scratch_bucket_creator.go

```diff
@@ -45,38 +46,50 @@ func NewScratchBucketCreator(ctx context.Context, storageClient domain.StorageCl
 // CreateScratchBucket creates scratch bucket in the same region as sourceFileFlag.
 // Returns (bucket_name, region, error)
 func (c *ScratchBucketCreator) CreateScratchBucket(
-	sourceFileFlag string, project string) (string, string, error) {
-	bucket := ""
-	region := ""
-	var err error
+	sourceFileFlag string, project string, zone string) (string, string, error) {
 
 	if project == "" {
 		return "", "", fmt.Errorf("can't create scratch bucket if project not specified")
 	}
+
 	if sourceFileFlag != "" {
 		// source file provided, create bucket in the same region for cost/performance reasons
-		bucket, region, err = c.createBucketMatchFileRegion(sourceFileFlag, project)
+		return c.createBucketMatchFileRegion(sourceFileFlag, project, zone)
 	}
-	if sourceFileFlag == "" || err != nil {
-		// source file not provided or couldn't create bucket based on it, create default scratch bucket
-		bucket = c.formatScratchBucketName(project, defaultRegion)
-		region, err = c.createBucketIfNotExisting(bucket, project, &storage.BucketAttrs{Name: bucket, Location: defaultRegion, StorageClass: defaultStorageClass})
-		if err != nil {
+
+	// if source file is not provided, fallback to input / default region.
+	return c.createBucketOnFallbackZone(project, zone)
+}
+
+func (c *ScratchBucketCreator) createBucketOnFallbackZone(project string, zone string) (string, string, error) {
+	fallbackRegion := defaultRegion
+	storageClass := defaultStorageClass
+	var err error
+	if zone != "" {
+		if fallbackRegion, err = getRegion(zone); err != nil {
 			return "", "", err
 		}
+		storageClass = regionalStorageClass
+	}
+	bucket := c.formatScratchBucketName(project, fallbackRegion)
+	region, err := c.createBucketIfNotExisting(bucket, project, &storage.BucketAttrs{Name: bucket, Location: fallbackRegion, StorageClass: storageClass})
+	if err != nil {
+		return "", "", err
 	}
 	return bucket, region, nil
 }
 
-func (c *ScratchBucketCreator) createBucketMatchFileRegion(fileGcsPath string, project string) (string, string, error) {
+func (c *ScratchBucketCreator) createBucketMatchFileRegion(fileGcsPath string, project string, zone string) (string, string, error) {
```

Comment:
The logic for creating determining where to create a bucket can be simplified without introducing another method (and calling it from two places).

We have as inputs:
- source file (optional)
- zone (optional)

We need to figure out:
- storage region
- storage class

Once these two are determined, bucket can be created in a single spot in the code. Error can be returned along the way if necessary.

Reply:
Yes we can do it that way. It's clearer, but not significantly clearer due to the complexity. Check out my new version.